funcoes_atv_2.py

```diff
@@ -0,0 +1,156 @@
+    for i in range(3):
+        try:
+            quantity_inputed = input("digite a quantidade do produto, ou tecle enter para adicionar 1: ")
+            if quantity_inputed == "":
+                pass
+            elif quantity_inputed.isnumeric() is True:
+                quantity = quantity_inputed
+            else:
+                raise ValueError
+            break
+        except ValueError:
+            print("Digite um valor válido.")
+            if i == 2:
+                print("Número de tentativas excedido.")
+                return main()
+    shopping_list.update({products: quantity})
+    if input("Deseja adicionar mais produtos? (s/n): ") == "s":
+        return add_item()
+    return main()
+
+
+def remove_item():
+    """
+    Remove um item da lista de compras.
+    Se o produto não estiver na lista, o usuário terá a opção de tentar novamente.
+    Se em três tentativas o produto não for encontrado, o usuário será redirecionado para a função principal.
+    """
+    for i in range(3):
+        try:
+            product_to_remove = input("Digite o produto que deseja remover: ")
+            if product_to_remove not in shopping_list.keys():
+                raise ValueError
+            shopping_list.pop(product_to_remove)
+            print(f"Produto {product_to_remove} removido com sucesso.")
+            if input("Deseja remover mais produtos? (s/n): ") == "s":
+                return remove_item()
+            return main()
+        except ValueError:
+            if input("Produto não encontrado. Tentar novamente? (s/n): ") in ["n", "não"]:
+                return main()
+    print("numero de tentativas excedido.")
+    return main()
+
+
+def edit_quantity():
+    """
+    Edita a quantidade de um item na lista de compras.
+    Se o produto não estiver na lista, o usuário terá a opção de tentar novamente.
+    Se o usuário inserir uma quantidade inválida, ele terá três tentativas antes de retornar à função principal.
+    """
+    product_to_edit = input("Digite o produto que deseja editar a quantidade: ")
+    if product_to_edit not in shopping_list:
+        print("Produto não encontrado.")
+        if input("Deseja tentar novamente? (s/n): ").lower() in ["s", "sim"]:
+            return edit_quantity()
+        return main()
+
+    for i in range(3):
+        try:
+            new_quantity = int(input("Digite a nova quantidade: "))
+            shopping_list[product_to_edit] = new_quantity
+            print(f"Quantidade de {product_to_edit} atualizada para {new_quantity}")
+            if input("Deseja editar mais produtos? (s/n): ").lower() == "s":
+                return edit_quantity()
+            return main()
+        except ValueError:
+            print("Digite um valor inteiro.")
+            if i == 2:
+                print("Número de tentativas excedido.")
+                return main()
+    return main()
+
+
+def view_shopping_list(**kwargs):
```

Comment:
O jeito que você usou o `**kwargs` não está errado, mas normalmente usamos para outros contextos. Vou tentar explicar aqui via texto mas podemos falar depois.
Imagina que você vai configurar uma conexão com seu banco de dados. Tem alguns parâmetros que são obrigatórios e que a gente precisa passar (a uri do banco, o nome do banco, etc). Alguns outros parâmetros são opcionais mas ainda assim customizáveis (coisas internas do banco e menos comuns de se mexer). Daí, o código que faz a conexão com o banco normalmente vai usar o `kwargs` para isso.
```
def conecta_no_banco(uri, nome_do_banco):
    conexao = conecta(uri, nome_do_banco)
    ...
    path_do_log = kwargs.get('path_do_log', '/tmp/logs')
    ...
```

Nesse caso (que inventei pra exemplificar pra você), o kwargs é um dicionário opcional em que você pode passar inúmeras opções para configurar seu banco. Normalmente a gente faz isso para não ter que aceitar uma infinidade de parâmetros obrigatórios em uma função. Já imaginou ter que receber como parâmetro obrigatório todas as possíveis configurações de um banco de dados?

Então assim, o uso que você fez não está errado, mas normalmente usamos em outros contextos. Viajando no teu exemplo, a tua função deveria sempre exibir o que está em `shopping_list`. O `kwargs` poderia ter uma parâmetro que, por ex, poderia juntar a lista que você passar mais a sua lista em `shopping_list`.

Reply:
Só pra deixar claro, não é pra usar somente com banco de dados, ok? Tem a ver com a lógica em si e não necessariamente com o banco.